// vso-node-api/HttpClient.js
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const url = require("url");
const http = require("http");
const https = require("https");
const tunnel = require("tunnel");
http.globalAgent.maxSockets = 100;
var HttpCodes;
(function (HttpCodes) {
    HttpCodes[HttpCodes["OK"] = 200] = "OK";
    HttpCodes[HttpCodes["MultipleChoices"] = 300] = "MultipleChoices";
    HttpCodes[HttpCodes["MovedPermanantly"] = 301] = "MovedPermanantly";
    HttpCodes[HttpCodes["ResourceMoved"] = 302] = "ResourceMoved";
    HttpCodes[HttpCodes["NotModified"] = 304] = "NotModified";
    HttpCodes[HttpCodes["UseProxy"] = 305] = "UseProxy";
    HttpCodes[HttpCodes["SwitchProxy"] = 306] = "SwitchProxy";
    HttpCodes[HttpCodes["TemporaryRedirect"] = 307] = "TemporaryRedirect";
    HttpCodes[HttpCodes["PermanentRedirect"] = 308] = "PermanentRedirect";
    HttpCodes[HttpCodes["BadRequest"] = 400] = "BadRequest";
    HttpCodes[HttpCodes["Unauthorized"] = 401] = "Unauthorized";
    HttpCodes[HttpCodes["PaymentRequired"] = 402] = "PaymentRequired";
    HttpCodes[HttpCodes["Forbidden"] = 403] = "Forbidden";
    HttpCodes[HttpCodes["NotFound"] = 404] = "NotFound";
    HttpCodes[HttpCodes["MethodNotAllowed"] = 405] = "MethodNotAllowed";
    HttpCodes[HttpCodes["NotAcceptable"] = 406] = "NotAcceptable";
    HttpCodes[HttpCodes["ProxyAuthenticationRequired"] = 407] = "ProxyAuthenticationRequired";
    HttpCodes[HttpCodes["RequestTimeout"] = 408] = "RequestTimeout";
    HttpCodes[HttpCodes["Conflict"] = 409] = "Conflict";
    HttpCodes[HttpCodes["Gone"] = 410] = "Gone";
    HttpCodes[HttpCodes["InternalServerError"] = 500] = "InternalServerError";
    HttpCodes[HttpCodes["NotImplemented"] = 501] = "NotImplemented";
    HttpCodes[HttpCodes["BadGateway"] = 502] = "BadGateway";
    HttpCodes[HttpCodes["ServiceUnavailable"] = 503] = "ServiceUnavailable";
    HttpCodes[HttpCodes["GatewayTimeout"] = 504] = "GatewayTimeout";
})(HttpCodes = exports.HttpCodes || (exports.HttpCodes = {}));
class HttpClientResponse {
    constructor(message) {
        this.message = message;
    }
    readBody() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            let output = '';
            this.message.on('data', (chunk) => {
                output += chunk;
            });
            this.message.on('end', () => {
                resolve(output);
            });
        }));
    }
}
exports.HttpClientResponse = HttpClientResponse;
function isHttps(requestUrl) {
    let parsedUrl = url.parse(requestUrl);
    return parsedUrl.protocol === 'https:';
}
exports.isHttps = isHttps;
class HttpClient {
    constructor(userAgent, handlers, socketTimeout) {
        this.userAgent = userAgent;
        this.handlers = handlers;
        this.socketTimeout = socketTimeout ? socketTimeout : 3 * 60000;
    }
    get(requestUrl, additionalHeaders) {
        return this.request('GET', requestUrl, null, additionalHeaders || {});
    }
    del(requestUrl, additionalHeaders) {
        return this.request('DELETE', requestUrl, null, additionalHeaders || {});
    }
    post(requestUrl, data, additionalHeaders) {
        return this.request('POST', requestUrl, data, additionalHeaders || {});
    }
    patch(requestUrl, data, additionalHeaders) {
        return this.request('PATCH', requestUrl, data, additionalHeaders || {});
    }
    put(requestUrl, data, additionalHeaders) {
        return this.request('PUT', requestUrl, data, additionalHeaders || {});
    }
    sendStream(verb, requestUrl, stream, additionalHeaders) {
        return this.request(verb, requestUrl, stream, additionalHeaders);
    }
    /**
     * Makes a raw http request.
     * All other methods such as get, post, patch, and request ultimately call this.
     * Prefer get, del, post and patch
     */
    request(verb, requestUrl, data, headers) {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            try {
                var info = this._prepareRequest(verb, requestUrl, headers);
                let res = yield this._requestRaw(info, data);
                // TODO: check 401 if handled
                // TODO: retry support
                resolve(res);
            }
            catch (err) {
                // only throws in truly exceptional cases (connection, can't resolve etc...)
                // responses from the server do not throw
                reject(err);
            }
        }));
    }
    _requestRaw(info, data) {
        return new Promise((resolve, reject) => {
            let socket;
            let isDataString = typeof (data) === 'string';
            if (typeof (data) === 'string') {
                info.options.headers["Content-Length"] = Buffer.byteLength(data, 'utf8');
            }
            let req = info.httpModule.request(info.options, (msg) => {
                let res = new HttpClientResponse(msg);
                resolve(res);
            });
            req.on('socket', (sock) => {
                socket = sock;
            });
            // If we ever get disconnected, we want the socket to timeout eventually
            req.setTimeout(this.socketTimeout, () => {
                if (socket) {
                    socket.end();
                }
                reject(new Error('Request timeout: ' + info.options.path));
            });
            req.on('error', function (err) {
                // err has statusCode property
                // res should have headers
                reject(err);
            });
            if (data && typeof (data) === 'string') {
                req.write(data, 'utf8');
            }
            if (data && typeof (data) !== 'string') {
                data.on('close', function () {
                    req.end();
                });
                data.pipe(req);
            }
            else {
                req.end();
            }
        });
    }
    _prepareRequest(method, requestUrl, headers) {
        let info = {};
        info.parsedUrl = url.parse(requestUrl);
        let usingSsl = info.parsedUrl.protocol === 'https:';
        info.httpModule = usingSsl ? https : http;
        var defaultPort = usingSsl ? 443 : 80;
        var proxyUrl;
        if (process.env.HTTPS_PROXY && usingSsl) {
            proxyUrl = url.parse(process.env.HTTPS_PROXY);
        }
        else if (process.env.HTTP_PROXY) {
            proxyUrl = url.parse(process.env.HTTP_PROXY);
        }
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
        info.options = {};
        info.options.host = info.parsedUrl.hostname;
        info.options.port = info.parsedUrl.port ? parseInt(info.parsedUrl.port) : defaultPort;
        info.options.path = (info.parsedUrl.pathname || '') + (info.parsedUrl.search || '');
        info.options.method = method;
        info.options.headers = headers || {};
        info.options.headers["User-Agent"] = this.userAgent;
        let useProxy = proxyUrl && proxyUrl.hostname;
        if (useProxy) {
            var agentOptions = {
                maxSockets: http.globalAgent.maxSockets,
                proxy: {
                    // TODO: support proxy-authorization
                    //proxyAuth: "user:password",
                    host: proxyUrl.hostname,
                    port: proxyUrl.port
                }
            };
            var tunnelAgent;
            var overHttps = proxyUrl.protocol === 'https:';
            if (usingSsl) {
                tunnelAgent = overHttps ? tunnel.httpsOverHttps : tunnel.httpsOverHttp;
            }
            else {
                tunnelAgent = overHttps ? tunnel.httpOverHttps : tunnel.httpOverHttp;
            }
            info.options.agent = tunnelAgent(agentOptions);
        }
        // gives handlers an opportunity to participate
        if (this.handlers) {
            this.handlers.forEach((handler) => {
                handler.prepareRequest(info.options);
            });
        }
        return info;
    }
}
exports.HttpClient = HttpClient;
//
// Legacy callback client.  Will delete.
//
class HttpCallbackClient {
    constructor(userAgent, handlers, socketTimeout) {
        this.userAgent = userAgent;
        this.handlers = handlers;
        this.socketTimeout = socketTimeout ? socketTimeout : 3 * 60000;
    }
    get(verb, requestUrl, headers, onResult) {
        var options = this._getOptions(verb, requestUrl, headers);
        this.request(options.protocol, options.options, null, onResult);
    }
    // POST, PATCH, PUT
    send(verb, requestUrl, data, headers, onResult) {
        var options = this._getOptions(verb, requestUrl, headers);
        this.request(options.protocol, options.options, data, onResult);
    }
    sendStream(verb, requestUrl, stream, headers, onResult) {
        var options = this._getOptions(verb, requestUrl, headers);
        var req = options.protocol.request(options.options, (res) => {
            let output = '';
            res.on('data', function (chunk) {
                output += chunk;
            });
            res.on('end', function () {
                // res has statusCode and headers
                onResult(null, res, output);
            });
        });
        req.on('error', function (err) {
            // err has statusCode property
            // res should have headers
            onResult(err, null, null);
        });
        stream.on('close', function () {
            req.end();
        });
        stream.pipe(req);
    }
    getStream(requestUrl, accept, onResult) {
        let headers = {};
        headers['ACCEPT'] = accept;
        var options = this._getOptions('GET', requestUrl, headers);
        var req = options.protocol.request(options.options, (res) => {
            onResult(null, res.statusCode, res);
        });
        req.on('error', (err) => {
            onResult(err, err.statusCode, null);
        });
        req.end();
    }
    /**
     * Makes an http request delegating authentication to handlers.
     * returns http result as contents buffer
     * All other methods such as get, post, and patch ultimately call this.
     */
    request(protocol, options, data, onResult) {
        // Set up a callback to pass off 401s to an authentication handler that can deal with it
        var callback = (err, res, contents) => {
            var authHandler;
            if (this.handlers) {
                this.handlers.some(function (handler, index, handlers) {
                    // Find the first one that can handle the auth based on the response
                    if (handler.canHandleAuthentication(res)) {
                        authHandler = handler;
                        return true;
                    }
                    return false;
                });
            }
            if (authHandler !== undefined) {
                authHandler.handleAuthentication(this, protocol, options, data, onResult);
            }
            else {
                // No auth handler found, call onResult normally
                onResult(err, res, contents);
            }
        };
        this.requestRaw(protocol, options, data, callback);
    }
    /**
     * Makes a raw http request.
     * All other methods such as get, post, patch, and request ultimately call this.
     */
    requestRaw(protocol, options, data, onResult) {
        var socket;
        if (data) {
            options.headers["Content-Length"] = Buffer.byteLength(data, 'utf8');
        }
        var callbackCalled = false;
        var handleResult = (err, res, contents) => {
            if (!callbackCalled) {
                callbackCalled = true;
                onResult(err, res, contents);
            }
        };
        var req = protocol.request(options, (res) => {
            let output = '';
            res.setEncoding('utf8');
            res.on('data', (chunk) => {
                output += chunk;
            });
            res.on('end', () => {
                // res has statusCode and headers
                handleResult(null, res, output);
            });
        });
        req.on('socket', (sock) => {
            socket = sock;
        });
        // If we ever get disconnected, we want the socket to timeout eventually
        req.setTimeout(this.socketTimeout, function () {
            if (socket) {
                socket.end();
            }
            handleResult(new Error('Request timeout: ' + options.path), null, null);
        });
        req.on('error', function (err) {
            // err has statusCode property
            // res should have headers
            handleResult(err, null, null);
        });
        if (data) {
            req.write(data, 'utf8');
        }
        req.end();
    }
    _getOptions(method, requestUrl, headers) {
        var parsedUrl = url.parse(requestUrl);
        var usingSsl = parsedUrl.protocol === 'https:';
        var prot = usingSsl ? https : http;
        var defaultPort = usingSsl ? 443 : 80;
        this.isSsl = usingSsl;
        var proxyUrl;
        if (process.env.HTTPS_PROXY && usingSsl) {
            proxyUrl = url.parse(process.env.HTTPS_PROXY);
        }
        else if (process.env.HTTP_PROXY) {
            proxyUrl = url.parse(process.env.HTTP_PROXY);
        }
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
        var options = {
            host: parsedUrl.hostname,
            port: parsedUrl.port || defaultPort,
            path: (parsedUrl.pathname || '') + (parsedUrl.search || ''),
            method: method,
            headers: headers || {}
        };
        options.headers["User-Agent"] = this.userAgent;
        var useProxy = proxyUrl && proxyUrl.hostname;
        if (useProxy) {
            var agentOptions = {
                maxSockets: http.globalAgent.maxSockets,
                proxy: {
                    // TODO: support proxy-authorization
                    //proxyAuth: "user:password",
                    host: proxyUrl.hostname,
                    port: proxyUrl.port
                }
            };
            var tunnelAgent;
            var overHttps = proxyUrl.protocol === 'https:';
            if (usingSsl) {
                tunnelAgent = overHttps ? tunnel.httpsOverHttps : tunnel.httpsOverHttp;
            }
            else {
                tunnelAgent = overHttps ? tunnel.httpOverHttps : tunnel.httpOverHttp;
            }
            options.agent = tunnelAgent(agentOptions);
        }
        if (this.handlers) {
            this.handlers.forEach((handler) => {
                handler.prepareRequest(options);
            });
        }
        return {
            protocol: prot,
            options: options,
        };
    }
}
exports.HttpCallbackClient = HttpCallbackClient;
