// import-cost/dist/src/utils.js
'use strict';

var fs = require('fs');
var path = require('path');
var pkgDir = require('pkg-dir');

function parseJson(dir) {
  var pkg = path.join(dir, 'package.json');
  return JSON.parse(fs.readFileSync(pkg, 'utf-8'));
}

function getPackageName(pkg) {
  var pkgParts = pkg.name.split('/');
  var pkgName = pkgParts.shift();
  if (pkgName.startsWith('@')) {
    pkgName = path.join(pkgName, pkgParts.shift());
  }
  return pkgName;
}

/**
 * @param {Object} pkg
 * @returns {string} The location of a node_modules folder containing this package.
 */
function getPackageModuleContainer(pkg) {
  var currentDir = path.dirname(pkg.fileName);
  var foundDir = '';
  var pkgName = getPackageName(pkg);

  while (!foundDir) {
    var projectDir = pkgDir.sync(currentDir);
    if (!projectDir) {
      throw new Error('Package directory not found [' + pkg.name + ']');
    }
    var modulesDirectory = path.join(projectDir, 'node_modules');
    if (fs.existsSync(path.resolve(modulesDirectory, pkgName))) {
      foundDir = modulesDirectory;
    } else {
      currentDir = path.resolve(currentDir, '..');
    }
  }
  return foundDir;
}

/**
 * @param {Object} pkg
 * @returns {string} The actual location on-disk for this package.
 */
function getPackageDirectory(pkg) {
  var pkgName = getPackageName(pkg);

  var tmp = getPackageModuleContainer(pkg);
  return path.resolve(tmp, pkgName);
}

function getPackageVersion(pkg) {
  return getPackageName(pkg) + '@' + getPackageJson(pkg).version;
}

function getPackageJson(pkg) {
  return parseJson(getPackageDirectory(pkg));
}

module.exports = {
  getPackageJson: getPackageJson,
  getPackageModuleContainer: getPackageModuleContainer,
  getPackageDirectory: getPackageDirectory,
  getPackageVersion: getPackageVersion,
  parseJson: parseJson
};
