// vso-node-api/interfaces/ChatInterfaces.js
/*
 * ---------------------------------------------------------
 * Copyright(C) Microsoft Corporation. All rights reserved.
 * ---------------------------------------------------------
 *
 * ---------------------------------------------------------
 * Generated file, DO NOT EDIT
 * ---------------------------------------------------------
 */
"use strict";
var MessageType;
(function (MessageType) {
    /**
     * A normal message that is sent by a user in plain text
     */
    MessageType[MessageType["Normal"] = 0] = "Normal";
    /**
     * A system message like User Entered and User Left Room
     */
    MessageType[MessageType["System"] = 1] = "System";
    /**
     * A notification from a TFS event
     */
    MessageType[MessageType["Notification"] = 2] = "Notification";
    /**
     * An error messsage.
     */
    MessageType[MessageType["Error"] = 3] = "Error";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
exports.TypeInfo = {
    Message: {},
    MessageType: {
        enumValues: {
            "normal": 0,
            "system": 1,
            "notification": 2,
            "error": 3
        }
    },
    Room: {},
    User: {},
};
exports.TypeInfo.Message.fields = {
    messageType: {
        enumType: exports.TypeInfo.MessageType
    },
    postedTime: {
        isDate: true,
    },
};
exports.TypeInfo.Room.fields = {
    createdDate: {
        isDate: true,
    },
    lastActivity: {
        isDate: true,
    },
};
exports.TypeInfo.User.fields = {
    joinedDate: {
        isDate: true,
    },
    lastActivity: {
        isDate: true,
    },
};
