// vso-node-api/interfaces/OrganizationInterfaces.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
var CollectionStatus;
(function (CollectionStatus) {
    CollectionStatus[CollectionStatus["Unknown"] = 0] = "Unknown";
    CollectionStatus[CollectionStatus["Initial"] = 10] = "Initial";
    CollectionStatus[CollectionStatus["Enabled"] = 20] = "Enabled";
    CollectionStatus[CollectionStatus["LogicallyDeleted"] = 30] = "LogicallyDeleted";
    CollectionStatus[CollectionStatus["MarkedForPhysicalDelete"] = 40] = "MarkedForPhysicalDelete";
})(CollectionStatus = exports.CollectionStatus || (exports.CollectionStatus = {}));
var OrganizationSearchKind;
(function (OrganizationSearchKind) {
    OrganizationSearchKind[OrganizationSearchKind["Unknown"] = 0] = "Unknown";
    OrganizationSearchKind[OrganizationSearchKind["ById"] = 1] = "ById";
    OrganizationSearchKind[OrganizationSearchKind["ByName"] = 2] = "ByName";
    OrganizationSearchKind[OrganizationSearchKind["ByTenantId"] = 3] = "ByTenantId";
})(OrganizationSearchKind = exports.OrganizationSearchKind || (exports.OrganizationSearchKind = {}));
var OrganizationStatus;
(function (OrganizationStatus) {
    OrganizationStatus[OrganizationStatus["Unknown"] = 0] = "Unknown";
    OrganizationStatus[OrganizationStatus["Initial"] = 10] = "Initial";
    OrganizationStatus[OrganizationStatus["Enabled"] = 20] = "Enabled";
    OrganizationStatus[OrganizationStatus["MarkedForDelete"] = 30] = "MarkedForDelete";
})(OrganizationStatus = exports.OrganizationStatus || (exports.OrganizationStatus = {}));
var OrganizationType;
(function (OrganizationType) {
    OrganizationType[OrganizationType["Unknown"] = 0] = "Unknown";
    OrganizationType[OrganizationType["Personal"] = 1] = "Personal";
    OrganizationType[OrganizationType["Work"] = 2] = "Work";
})(OrganizationType = exports.OrganizationType || (exports.OrganizationType = {}));
exports.TypeInfo = {
    Collection: {
        fields: null
    },
    CollectionStatus: {
        enumValues: {
            "unknown": 0,
            "initial": 10,
            "enabled": 20,
            "logicallyDeleted": 30,
            "markedForPhysicalDelete": 40,
        }
    },
    Organization: {
        fields: null
    },
    OrganizationMigrationBlob: {
        fields: null
    },
    OrganizationSearchKind: {
        enumValues: {
            "unknown": 0,
            "byId": 1,
            "byName": 2,
            "byTenantId": 3,
        }
    },
    OrganizationStatus: {
        enumValues: {
            "unknown": 0,
            "initial": 10,
            "enabled": 20,
            "markedForDelete": 30,
        }
    },
    OrganizationType: {
        enumValues: {
            "unknown": 0,
            "personal": 1,
            "work": 2,
        }
    },
    Policy: {
        fields: null
    },
    PolicyInfo: {
        fields: null
    },
    Region: {
        fields: null
    },
};
exports.TypeInfo.Collection.fields = {
    dateCreated: {
        isDate: true,
    },
    lastUpdated: {
        isDate: true,
    },
    status: {
        enumType: exports.TypeInfo.CollectionStatus
    },
};
exports.TypeInfo.Organization.fields = {
    collections: {
        isArray: true,
        typeInfo: exports.TypeInfo.Collection
    },
    dateCreated: {
        isDate: true,
    },
    lastUpdated: {
        isDate: true,
    },
    primaryCollection: {
        typeInfo: exports.TypeInfo.Collection
    },
    status: {
        enumType: exports.TypeInfo.OrganizationStatus
    },
    type: {
        enumType: exports.TypeInfo.OrganizationType
    },
};
exports.TypeInfo.OrganizationMigrationBlob.fields = {};
exports.TypeInfo.Policy.fields = {
    parentPolicy: {
        typeInfo: exports.TypeInfo.Policy
    },
};
exports.TypeInfo.PolicyInfo.fields = {};
exports.TypeInfo.Region.fields = {};
