// vso-node-api/interfaces/PolicyInterfaces.js
/*
 * ---------------------------------------------------------
 * Copyright(C) Microsoft Corporation. All rights reserved.
 * ---------------------------------------------------------
 *
 * ---------------------------------------------------------
 * Generated file, DO NOT EDIT
 * ---------------------------------------------------------
 */
"use strict";
var PolicyEvaluationStatus;
(function (PolicyEvaluationStatus) {
    PolicyEvaluationStatus[PolicyEvaluationStatus["Queued"] = 0] = "Queued";
    PolicyEvaluationStatus[PolicyEvaluationStatus["Running"] = 1] = "Running";
    PolicyEvaluationStatus[PolicyEvaluationStatus["Approved"] = 2] = "Approved";
    PolicyEvaluationStatus[PolicyEvaluationStatus["Rejected"] = 3] = "Rejected";
    PolicyEvaluationStatus[PolicyEvaluationStatus["NotApplicable"] = 4] = "NotApplicable";
    PolicyEvaluationStatus[PolicyEvaluationStatus["Broken"] = 5] = "Broken";
})(PolicyEvaluationStatus = exports.PolicyEvaluationStatus || (exports.PolicyEvaluationStatus = {}));
exports.TypeInfo = {
    PolicyConfiguration: {},
    PolicyEvaluationRecord: {},
    PolicyEvaluationStatus: {
        enumValues: {
            "queued": 0,
            "running": 1,
            "approved": 2,
            "rejected": 3,
            "notApplicable": 4,
            "broken": 5
        }
    },
};
exports.TypeInfo.PolicyConfiguration.fields = {
    createdDate: {
        isDate: true,
    },
};
exports.TypeInfo.PolicyEvaluationRecord.fields = {
    completedDate: {
        isDate: true,
    },
    configuration: {
        typeInfo: exports.TypeInfo.PolicyConfiguration
    },
    startedDate: {
        isDate: true,
    },
    status: {
        enumType: exports.TypeInfo.PolicyEvaluationStatus
    },
};
